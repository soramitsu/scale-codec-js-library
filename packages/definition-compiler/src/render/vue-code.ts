/* eslint-disable max-params */

import { App, createRenderer, nextTick } from 'vue';
import prettier from 'prettier';

type CodeNode = CodeElement;

class CodeElement {
    public parent: null | CodeNode = null;
    public name: string | null = null;
    public props: any = {};
    public text?: string;
    public comment?: string;
    public children: CodeNode[] = [];

    public constructor(params: Partial<CodeElement>) {
        Object.assign(this, params);
    }

    public render(): string {
        if (this.name === 'text' && this.text) {
            return this.text;
        }
        return this.children.map((x) => x.render()).join('');
    }
}

export const { createApp } = createRenderer<CodeNode, CodeElement>({
    // create
    createElement(type, isSVG, isCustomized, props) {
        return new CodeElement({
            name: type,
            props,
        });
    },
    createComment(comment) {
        return new CodeElement({
            name: 'comment',
            comment,
        });
    },
    createText(text) {
        return new CodeElement({
            name: 'text',
            text,
        });
    },
    setElementText(elem, txt) {
        elem.text = txt;
    },
    insert(el, parent, anchor) {
        // console.log('insert %o', { parent, el, anchor });
        el.parent = parent;
        if (anchor) {
            const index = parent.children.indexOf(anchor);
            if (index >= 0) {
                parent.children.splice(index, 0, el);
                return;
            }
        }
        parent.children.push(el);
    },
    patchProp(el, key, prev, next, isSVG, prevChildren, parentComponent, parentSuspense, unmountChildren) {
        el.props[key] = next;
    },
    remove(el) {
        if (el.parent) {
            el.parent.children.splice(el.parent.children.indexOf(el), 1);
            el.parent = null;
        }
    },
    nextSibling(node) {
        // console.log('next', { node });
        if (node.parent) {
            const index = node.parent.children.indexOf(node);
            return node.parent.children[index + 1] ?? null;
        }
        return null;
    },
    setText(node, text) {
        node.text = text;
    },
    parentNode(node) {
        return node.parent;
    },
});

function renderNode(node: CodeNode): string {
    const rendered = node.render();

    const prettified = prettier.format(rendered, {
        semi: false,
        tabWidth: 2,
        singleQuote: true,
        printWidth: 120,
        parser: 'typescript',
    });

    return prettified;
}

export async function renderApp(app: App<CodeElement>): Promise<string> {
    const root: CodeNode = new CodeElement({ name: 'root', children: [] });
    app.config.compilerOptions.whitespace = 'preserve';
    app.mount(root);

    await nextTick();

    return renderNode(root);
}
