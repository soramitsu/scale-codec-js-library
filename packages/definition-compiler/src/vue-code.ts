/* eslint-disable max-params */

import { App, createRenderer } from 'vue';
import prettier from 'prettier';

type CodeNode = CodeElement;

class CodeElement {
    public parent: null | CodeNode = null;
    public name: string | null = null;
    public props: any = {};
    public text?: string;
    public comment?: string;
    public children: CodeNode[] = [];

    public constructor(params: Partial<CodeElement>) {
        Object.assign(this, params);
    }

    public render(): string {
        if (this.name === 'text' && this.text) {
            return this.text;
        }
        return this.children.map((x) => x.render()).join('');
    }
}

export const { createApp } = createRenderer<CodeNode, CodeElement>({
    // create
    createElement(type, isSVG, isCustomized, props) {
        return new CodeElement({
            name: type,
            props,
        });
    },
    createComment(comment) {
        return new CodeElement({
            name: 'comment',
            comment,
        });
    },
    createText(text) {
        return new CodeElement({
            name: 'text',
            text,
        });
    },
    setElementText(elem, txt) {
        elem.text = txt;
    },
    insert(el, parent, anchor) {
        if (anchor) {
            const index = parent.children.indexOf(anchor);
            if (index >= 0) {
                parent.children.splice(index, 0, el);
                return;
            }
        }
        el.parent = parent;
        parent.children.push(el);
    },
    patchProp(el, key, prev, next, isSVG, prevChildren, parentComponent, parentSuspense, unmountChildren) {
        el.props[key] = next;
    },
    remove(el) {
        if (el.parent) {
            el.parent.children.splice(el.parent.children.indexOf(el), 1);
        }
    },
    nextSibling(node) {
        if (node.parent) {
            const index = node.parent.children.indexOf(node);
            return node.parent.children[index + 1] ?? null;
        }
        return null;
    },
    setText(node, text) {
        node.text = text;
    },
    parentNode(node) {
        return node.parent;
    },
});

function renderNode(node: CodeNode): string {
    const rendered = node.render();

    const prettified = prettier.format(rendered, {
        semi: false,
        tabWidth: 2,
        parser: 'typescript',
    });

    return prettified;
}

export function renderApp(app: App<CodeElement>): string {
    const root: CodeNode = new CodeElement({ name: 'root', children: [] });
    app.config.compilerOptions.whitespace = 'preserve';
    app.mount(root);
    return renderNode(root);
}
