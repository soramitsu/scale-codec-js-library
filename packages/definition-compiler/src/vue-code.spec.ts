import { defineComponent, compile } from 'vue';
import { createApp, renderApp } from './vue-code';

it('renders just text', () => {
    const app = createApp({
        render: compile(`just text`),
    });

    expect(renderApp(app)).toEqual('just text');
});

it.only('renders child components', () => {
    const Child = defineComponent({
        props: {
            msg: String,
        },
        render: compile(`Message - {{ msg }}`),
    });

    const Ws = defineComponent({
        props: {
            text: String,
        },
        render: ({ text }) => text,
    });

    const app = createApp({
        components: { Child, Ws },
        render: compile(`
            <Child msg="nya" />
            <Ws text=" | " />
            <Child msg="Nya?" />
        `),
    });

    expect(renderApp(app)).toEqual('Message - nya | Message - Nya?');
});
