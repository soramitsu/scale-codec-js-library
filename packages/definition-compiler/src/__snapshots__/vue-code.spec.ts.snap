// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders child components 1`] = `""`;

exports[`renders just text 1`] = `""`;
